src/librustc_typeck/check/coercion.rs

```diff
@@ -1244,26 +1243,46 @@ impl<'gcx, 'tcx, 'exprs, E> CoerceMany<'gcx, 'tcx, 'exprs, E>
                         // as prior return coercions would not be relevant (#57664).
                         let parent_id = fcx.tcx.hir().get_parent_node(blk_id);
                         let parent = fcx.tcx.hir().get(fcx.tcx.hir().get_parent_node(parent_id));
```

Comment:
You can remove the `get` call and use `get_fn_decl` instead of `get_node_fn_decl`. After that change it looks to me like this code and the almost-copy below can be merged (maybe into a function called twice)

Reply:
Done. Had to keep `get_fn_decl` and `get_node_fn_decl` separate because they have slightly different behavior that affects the output.